### PR TITLE
optimize softmax with cross entropy soft label

### PR DESCRIPTION
**Pull request by xingfeng01**
### PR types
Performance optimization

### PR changes
OPs

### Describe
softmax_with_cross_entropy optimization with soft label. This PR includes optimization of 
- "SoftmaxWithCrossEntropySoftLabel" : compute log_softmax and then compute loss. 
- "CrossEntropySoftLabel" : compute loss with softmax as input.

These optimization includes following technics:
- read data to buffer with vectorization
- compute max and sum in warp
- fixed loop size with macro

Performance (computation time):
- softmax_with_cross_entropy_0 (forward) : -40.1%
- softmax_with_cross_entropy_0 (backward): -41%

### CODE REVIEW

**Comment by xingfeng01** [comment]
Performance optimized, computation time reduced:

softmax_with_cross_entropy_0 (forward) : -40.1%
softmax_with_cross_entropy_0 (backward): -41%

File: paddle/fluid/operators/softmax_with_cross_entropy_op.cu [hunk 1]
```diff
@@ -152,10 +152,12 @@ __global__ void CrossEntropyExpHardLabel(T* loss, T* softmax,
 */
 template <typename T, typename VecT, typename AccT, int Log2Elements,
           SoftmaxMode mode, bool IgnoreIndex>
-__global__ void WarpSoftmaxForward(T* loss, T* softmax, const T* src,
-                                   const int64_t* label, const int batch_size,
-                                   const int stride, const int element_count,
-                                   const int ignore_index) {
+__global__ void WarpSoftmaxForwardHardLabel(T* loss, T* softmax, const T* src,
```
Comment (Xreki) [inline paddle/fluid/operators/softmax_with_cross_entropy_op.cu]: 这个函数，按当初的设计，是即可用于计算softmax和log_softmax，又可用于计算softmax_with_cross_entropy（硬标签），所以可不必改名？

File: paddle/fluid/operators/softmax_with_cross_entropy_op.cu [hunk 2]
```diff
@@ -494,6 +498,356 @@ __global__ void SoftmaxWithCrossEntropyGradHardLabel(
...
 }
 
+/*
+  Cross entropy soft label with dynamic size on axis (Log2Elements is varibale).
+*/
+template <typename T, typename VecT, bool InLogMode = false>
+__global__ void CrossEntropySoftLabel(T* loss, T* softmaxwrt, const T* softmax,
+                                      const T* labels, const int n,
+                                      const int dim, const int d,
+                                      int Log2Elements) {
```
Comment (Xreki) [inline paddle/fluid/operators/softmax_with_cross_entropy_op.cu]: - `Log2Elements`不作为模板，而是作为输入参数传入，则应该命名为`log2_elements`。
- 这个函数模板中不必传`InLogMode`？
- `CrossEntropySoftLabel`和`WarpSoftmaxForwardSoftLabel`的差别是什么？代码结构看起来差不多。
- 输入参数`softmaxwrt`指的是什么？
- 调用处第三个参数传的`NULL`，那可以删掉？
- 调用处`T`和`VecT`都传的T，即`VecSize`为1，后续打算支持向量化吗？若不打算支持，则模板中可以删掉`VecT`。

以上是我看到这个函数产生的一些问题。
----

通过对照文件中的全部实现和该函数调用，我大体明白了这个函数要支持2种情况：

- 输入是softmax，只需使用该softmax计算loss
- 输入是log_softmax，需要用log_softmax计算loss，并更新softmax。

建议函数命名、输入参数定义的更清晰一些。另外，给函数加一些解释说明。

File: paddle/fluid/operators/softmax_with_cross_entropy_op.cu [hunk 3]
```diff
@@ -494,6 +498,356 @@ __global__ void SoftmaxWithCrossEntropyGradHardLabel(
...
+    PADDLE_ENFORCE_CUDA_SUCCESS(platform::dynload::cudnnSoftmaxForward(
+        handle, CUDNN_SOFTMAX_LOG, mode, platform::CudnnDataType<T>::kOne(),
+        descp, logits_data, platform::CudnnDataType<T>::kZero(), descp,
+        softmax_data));
+#endif
+
+    const int kDimLog2 = static_cast<int>(Log2Ceil(dim));
+    const int kDimCeil = 1 << kDimLog2;
+
+    int kThreadPerBlock = 512;
```
Comment (Xreki) [inline paddle/fluid/operators/softmax_with_cross_entropy_op.cu:848]: 安全起见，rocm平台的block_size不要超过256。

File: paddle/fluid/operators/softmax_with_cross_entropy_op.cu [hunk 4]
```diff
@@ -494,6 +498,356 @@ __global__ void SoftmaxWithCrossEntropyGradHardLabel(
...
+    SOFTMAX_WARP_FORWARD_SOFT_CASE(7, T, AccT);
+    SOFTMAX_WARP_FORWARD_SOFT_CASE(8, T, AccT);
+    SOFTMAX_WARP_FORWARD_SOFT_CASE(9, T, AccT);
+    default:
+      break;
+  }
+}
+
+template <typename T>
+static void SoftmaxWithCrossEntropySoftLabel(
```
Comment (Xreki) [inline paddle/fluid/operators/softmax_with_cross_entropy_op.cu:780]: 这个函数和`SoftmaxWithCrossEntropyHardLabel`相似度太高，建议合并成1个。

File: paddle/fluid/operators/softmax_with_cross_entropy_op.cu [hunk 5]
```diff
@@ -494,6 +498,356 @@ __global__ void SoftmaxWithCrossEntropyGradHardLabel(
...
+    int idx_n = ids / d;
+    int idx_d = ids % d;
+#pragma unroll
+    for (int it = 0; it < kIterations; ++it) {
+      int idx_dim = it * kThreadPerBatch + threadIdx.x;
+      int idx = idx_n * dim * d + idx_dim * d + idx_d;
+
+      if (idx_n < n && idx_dim < dim) {
+        VecT softmaxdata;
+        if (InLogMode == true) {
```
Comment (Xreki) [inline paddle/fluid/operators/softmax_with_cross_entropy_op.cu]: 直接用`if (InLogMode)`即可。下同。

File: paddle/fluid/operators/softmax_with_cross_entropy_op.cu [hunk 6]
```diff
@@ -494,6 +498,356 @@ __global__ void SoftmaxWithCrossEntropyGradHardLabel(
...
+          softmaxdata = reinterpret_cast<VecT*>(&softmaxwrt[idx])[0];
+        } else {
+          softmaxdata = reinterpret_cast<const VecT*>(&softmax[idx])[0];
+        }
+        VecT labelsdata = reinterpret_cast<const VecT*>(&labels[idx])[0];
+        T* softmaxptr = reinterpret_cast<T*>(&softmaxdata);
+        T* labelsptr = reinterpret_cast<T*>(&labelsdata);
```
Comment (Xreki) [inline paddle/fluid/operators/softmax_with_cross_entropy_op.cu:550]:
```cpp
template <typename T, int Size>
struct alignas(sizeof(T) * Size) AlignedVector {
  T val[Size];
};
```

感觉上述定义的`VecT`会更好用一些，避免指针使用`reinterpret_cast`转来转去。

**Review by Xreki — COMMENTED** [review]
很早之前的一些review建议，一直没有提交，作为参考。

File: paddle/fluid/operators/softmax_with_cross_entropy_op.cu (hunk 1 above)
Comment (xingfeng01) [reply paddle/fluid/operators/softmax_with_cross_entropy_op.cu]: 已修改

File: paddle/fluid/operators/softmax_with_cross_entropy_op.cu (hunk 2 above)
Comment (xingfeng01) [reply paddle/fluid/operators/softmax_with_cross_entropy_op.cu]: 已加注释

File: paddle/fluid/operators/softmax_with_cross_entropy_op.cu (hunk 3 above)
Comment (xingfeng01) [reply paddle/fluid/operators/softmax_with_cross_entropy_op.cu:848]: 所有 512 均已修改

File: paddle/fluid/operators/softmax_with_cross_entropy_op.cu (hunk 5 above)
Comment (xingfeng01) [reply paddle/fluid/operators/softmax_with_cross_entropy_op.cu]: 已修改

File: paddle/fluid/operators/softmax_with_cross_entropy_op.cu (hunk 4 above)
Comment (xingfeng01) [reply paddle/fluid/operators/softmax_with_cross_entropy_op.cu:780]: 好的，下个 PR 处理代码清理问题

File: paddle/fluid/operators/softmax_with_cross_entropy_op.cu (hunk 6 above)
Comment (xingfeng01) [reply paddle/fluid/operators/softmax_with_cross_entropy_op.cu:550]: 个人觉得目前的方式可读性高，比较清楚的看到做了cast操作

**Review by JiaXiao243 — DISMISSED** [review]
LGTM.  mv3_large_x1_0_distill_mv3_small_x1_0  top1 acc is 67.9%.
0fb6547764967cb0ee494e03fea9e6dc

**Review by yghstill — APPROVED** [review]
LGTM

**Review by lanxianghit — APPROVED** [review]
